src/core/thread/mle.cpp

```diff
@@ -1673,6 +1673,7 @@ otError Mle::SendChildIdRequest(void)
     SuccessOrExit(error = AppendPendingTimestamp(*message));
 
     mParentCandidate.SetState(Neighbor::kStateValid);
+    SetRloc16(Mac::kShortAddrInvalid);
```

Comment:
I don't think we should update the RLOC16 here.  If the device fails to complete the attach by receiving a Child ID Response message, it should continue with its existing parent.

Instead, how about changing the MAC source address selection logic at [mesh_forwarder.cpp#L626](https://github.com/openthread/openthread/blob/master/src/core/thread/mesh_forwarder.cpp#L626)?  Specifically, use Extended Address whenever `GetParentCandidate()` does not return the same as `GetParent()`.

Reply:
Sounds good. 

Another (unrelated) thing is whether we should use the short or extended address for parent...Current code matches it with source (i.e., if src uses short we use short for dst, otherwise extended for both src and dst). I think we can always use short for dst. Thoughts?

Reply:
We need to use the Extended Address in DST to handle the case where the Parent Candidate is not yet a router.  The following occurs:
1) Parent Candidate is current a REED (i.e. child).
2) Parent Candidate receives Child ID Request.
3) Parent Candidate sends Address Solicit to Leader.
4) Parent Candidate receives Address Response from Leader and updates its RLOC16.
5) Parent Candidate queues Child ID Response.
6) Attaching child has no idea what the Parent Candidate's RLOC16 is.

The existing logic is simple.  We could optimize it to use the Short Address as DST if the Parent Candidate is already a router.

Reply:
Thanks. Let's keep the logic the same and simple.
I will close this PR and have the new changes in another PR.